src/main/java/Duke/Storage.java

```diff
@@ -0,0 +1,199 @@
+            return "D | " + (task.isDone() ? "1" : "0") + " | " + deadlineTask.getDescription() + " | " + "by " + by;
+        } else if (task instanceof Event) {
+            Event eventTask = (Event) task;
+            String from = eventTask.getFrom().format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+            String to = eventTask.getTo().format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+            return "E | " + (task.isDone() ? "1" : "0") + " | "
+                    + eventTask.getDescription() + " | " + from + " to " + to;
+        } else {
+            // Handle other task types if needed
+            return "";
+        }
+    }
+}
+
+/**
+ * The TaskListDecoder class decodes a list of strings from a file into a list of tasks.
+ */
+class TaskListDecoder {
+    /**
+     * Decodes a list of strings into a list of tasks.
+     *
+     * @param lines The list of strings to decode.
+     * @return The decoded list of tasks.
+     * @throws DukeDataCorruptedException If the data file is corrupted.
+     */
+    public static List<Task> decode(List<String> lines) throws DukeDataCorruptedException {
+        List<Task> decodedTasks = new ArrayList<>();
+        try {
+            for (String line : lines) {
+                decodedTasks.add(decodeTask(line));
+            }
+        } catch (IllegalArgumentException e) {
+            throw new DukeDataCorruptedException("Data file is corrupted: " + e.getMessage());
+        }
+
+        return decodedTasks;
+    }
+
+
+    private static Task decodeTask(String line) {
+        String[] parts = line.split(" \\| ");
+        String taskType = parts[0];
+        boolean isDone = parts[1].equals("1"); // Check if the task is marked as done
+        String description = parts[2];
+
+        switch (taskType) {
+        case "T":
+            Task newTodo = new Todo(description);
+            newTodo.setDone(isDone);
+            return newTodo;
+        case "D":
+            String by = parts[3];
+            LocalDate byDateTime = LocalDate.parse(by.substring(3), DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+            Task newDeadline = new Deadline(description, byDateTime);
+            newDeadline.setDone(isDone);
+            return newDeadline;
+        case "E":
+            String dateTimeString = parts[3];
+            String[] dateTimeParts = dateTimeString.split(" to ");
+            String from = dateTimeParts[0];
+            String to = dateTimeParts[1];
+
+            LocalDate fromDateTime = LocalDate.parse(from, DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+            LocalDate toDateTime = LocalDate.parse(to, DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+
+            Task newEvent = new Event(description, fromDateTime, toDateTime);
+            newEvent.setDone(isDone);
+            return newEvent;
+        default:
+            // Handle other task types if needed
+            return null;
```

Comment:
I like how well-organised the code is here with the clear separation of logical units